x/config/config.go

```diff
@@ -0,0 +1,193 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package config provides convenience functions to create [transport.StreamDialer] and [transport.PacketDialer]
+// objects based on a text config. This is experimental and mostly for illustrative purposes at this point.
+package config
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport/shadowsocks"
+	"github.com/Jigsaw-Code/outline-sdk/transport/socks5"
+	"github.com/Jigsaw-Code/outline-sdk/transport/split"
+)
+
+// NewStreamDialer creates a new [transport.StreamDialer] according to the given config.
+func NewStreamDialer(transportConfig string) (dialer transport.StreamDialer, err error) {
```

Comment:
I feel this can be put in a separate file (e.g. `stream.go`)? Also the `PacketDialer` related functions can be put to `packet.go`. Then this file will only contain `shadowsocksConfig` structure, `parseShadowsocksURL` and package doc.

Reply:
Thanks for the quick feedback. As mentioned privately, I believe grouping by protocol scales better, since they will share code. I moved the Shadowsocks code to its own file. I also found it helpful to see both the Stream and Packet factories in one place to compare what's supported where.